[WGSL] Overload resolution shouldn't assume we can materialize all variables

tadeuzagallo opened the pull request:
#### 9668cecef5520dd5a177d2fa0ad7af2a396f0010
<pre>
[WGSL] Overload resolution shouldn&apos;t assume we can materialize all variables
<a href="https://bugs.webkit.org/show_bug.cgi?id=254588">https://bugs.webkit.org/show_bug.cgi?id=254588</a>
rdar://problem/107316646

Reviewed by Myles C. Maxfield.

Some overload candidates require explicitly instantiating type variables, and
if no explicit instantiation is given we&apos;ll fail to materialize the variable.
In this case, we should reject the overload candidate instead of failing an
assertion and crashing.

* Source/WebGPU/WGSL/Overload.cpp:
(WGSL:: const):
(WGSL::OverloadResolver::considerCandidate):
* Source/WebGPU/WGSL/TypeDeclarations.rb:

Canonical link: <a href="https://commits.webkit.org/262323@main">https://commits.webkit.org/262323@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/68d69936128de5c6fcfbf908cdbb9a88bd347495

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1232 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1266 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1305 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2007 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1101 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1318 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1334 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/1262 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1241 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/1157 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/1151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1870 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/1181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1141 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/1115 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1126 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1172 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2225 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1175 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1080 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1111 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1151 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/307 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1201 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/Overload.cpp:
```diff
@@ -72,7 +72,7 @@ class OverloadResolver {
     void assign(TypeVariable, Type*);
     Type* resolve(TypeVariable) const;
-    Type* materialize(const AbstractType&) const;
+    std::optional<Type*> materialize(const AbstractType&) const;
```
Can't materializing a type create a new Type object? I don't quite understand how it only returns a raw pointer to a Type.

tadeuzagallo replied:
The lifecycle of types is handled by the TypeStore, which is owned by the ShaderModule. Types are hash-consed, so we produce the minimal set of types we'll need for the lifecycle of the program and keep they alive for as long as the module is alive.

litherum commented on Source/WebGPU/WGSL/Overload.cpp:
```diff
@@ -135,27 +135,31 @@ Type* OverloadResolver::resolve()
-            auto rows = materialize(matrix.rows);
-            return m_types.matrixType(element, columns, rows);
+        [&](const AbstractMatrix& matrix) -> std::optional<Type*> {
+            if (auto element = materialize(matrix.element)) {
```
I don't understand how `materialize(matrix.element)` knows that we're in the middle of materializing a matrix, so only some types are acceptable. There doesn't seem to be any connection between the nested materialize calls to provide context between them. 🤔

tadeuzagallo replied:
`AbstractMatrix` is only constructed from `TypeDeclarations.rb`, and the declaration provides constraints on the variable (if you look at all the matrix types, they all take variables constrained to `Float`). We materialize variables by looking them up in the type substitution dictionary, but we only insert a substitution if it satisfies the constraints, so by the time we get to materialization this has already been handled.

There's another part to this, which is validating user-provided types, which is not implemented yet (e.g. we should, but still don't, reject `let x : mat2x2<i32>;`). This will happen in `TypeCheck.cpp` though, so it's not part of overload resolution.

litherum commented on Source/WebGPU/WGSL/Overload.cpp:
```diff
@@ -216,7 +225,7 @@ std::optional<ViableOverload> OverloadResolver::considerCandidate(const Overload
                 dataLog(", ");
             first = false;
-            dataLog(*materialize(parameter));
+            dataLog(**materialize(parameter));
```
Optional pointer? That doesn't smell right. Pointers can already be optional 🤔

tadeuzagallo replied:
Oops, good call, not sure why I did that 🫣

tadeuzagallo committed:
[WGSL] Overload resolution shouldn't assume we can materialize all variables
https://bugs.webkit.org/show_bug.cgi?id=254588
rdar://problem/107316646

Reviewed by Myles C. Maxfield.

Some overload candidates require explicitly instantiating type variables, and
if no explicit instantiation is given we'll fail to materialize the variable.
In this case, we should reject the overload candidate instead of failing an
assertion and crashing.

* Source/WebGPU/WGSL/Overload.cpp:
(WGSL:: const):
(WGSL::OverloadResolver::considerCandidate):
* Source/WebGPU/WGSL/TypeDeclarations.rb:

Canonical link: https://commits.webkit.org/262323@main
